bundles/org.openhab.binding.goveelan/src/main/java/org/openhab/binding/goveelan/internal/GoveeLanHandler.java

```diff
@@ -0,0 +1,365 @@
+ * <li>The job that receives the answers and applies that to the respective thing is called <i>refreshStatusJob</i> and
+ * there is only one for all instances. It may be stopped and restarted by the DiscoveryService (see below).</li>
+ * </ul>
+ *
+ * The other topic that needs to be managed is that device discovery responses are also sent to openHAB at the same port
+ * like status updates. Therefore, when scanning new devices that job that listens to status devices must
+ * be stopped while scanning new devices. Otherwise, the status job will receive the scan discover UDB packages.
+ *
+ * @author Stefan Höhn - Initial contribution
+ */
+@NonNullByDefault
+public class GoveeLanHandler extends BaseThingHandler {
+
+    /*
+     * Messages to be sent to the Govee Devices
+     */
+    private static final String LIGHT_OFF = "{\"msg\": {\"cmd\": \"turn\", \"data\": {\"value\": \"0\"}}}";
+    // turning on via cmd-turn and value = 1 doesn't work, so let's use the brightness command
+    private static final String LIGHT_ON = """
+            {"msg": {"cmd": "brightness", "data": {"value": "100"}}}
+            """;
+    private static final String LIGHT_COLOR = """
+            {
+              "msg" : {
+                "cmd":"colorwc",
+                "data": {
+                  "color" : {
+                    "r" : %d,
+                    "g" : %d,
+                    "b":  %d
+                  },
+                  "colorTemInKelvin" : %d
+                }
+              }
+            }
+            """;
+    private static final String LIGHT_BRIGHTNESS = """
+            {
+              "msg":{
+                "cmd":"brightness",
+                "data":{
+                  "value": %d
+                }
+              }
+            }
+            """;
+
+    private static final String QUERY_STATUS = """
+            {
+              "msg":{
+                "cmd":"devStatus",
+                "data":{
+                }
+              }
+            }
+            """;
+
+    // Holds a list of all thing handlers to send them thing updates via the receiver-Thread
+    private static final Map<String, GoveeLanHandler> THING_HANDLERS = new HashMap<>();
+
+    private final Logger LOGGER = LoggerFactory.getLogger(GoveeLanHandler.class);
+    private static final int SENDTODEVICE_PORT = 4003;
+    private static final int RECEIVEFROMDEVICE_PORT = 4002;
+
+    // Semaphores to suppress further processing if already running
+    private static boolean refreshJobRunning = false;
+    private static boolean refreshRunning = false;
+
+    @Nullable
+    private static ScheduledFuture<?> refreshStatusJob; // device response receiver job
+    @Nullable
+    private ScheduledFuture<?> triggerStatusJob; // send device status update job
+
+    /*
+     * Common Receiver job for the status answers of the devices
+     */
+    public static boolean isRefreshJobRunning() {
+        return refreshJobRunning && THING_HANDLERS.isEmpty();
+    }
```

Comment:
It seems that concurrency is not correctly handled and there can be race conditions? In line 130 semaphores are mentioned, but I don't see any?

Reply:
This really becomes a pretty tricky topic I worked hours on. Concurrency is a nightmare here. 

Basically the two threads (the discovery in GoveeLanDiscoveryService and the status receiver in GoveeLanHandler have to listen on the same Multisocket-Port. Hence, I must not receive any status package while discovering and the other way around). Therefore Discovery needs to wait if currently a status-Receiver is accepting a new udb packet and then make not a new one is triggered. The "semaphores" are refreshJobRunning and refreshRunning. The latter one makes sure, only one refresh is triggered at a time and refreshJobRunning tells us if the handler is currently working on receiving a status. If that is the case, the discovery waits until the status-receivers receives a package and only then does the recovery. It then stops all Handler-status-receiver-jobs, so the discovery works without being disturbed by the handler and then after discovery the handler's status-Job is restarted again.

I've been through quite some daunting stuff with the Nanoleafs which is why I put a lot of work into this here. It only worked until I implemented it that way. I tested this quite intensively (as much as concurrency is testable) by really trying to break during discovery and receiving status and removing things and channel and at least I could not break it (before I had this implementation iteratively implemented it failed massively)

I hope that explains it.